torch/csrc/jit/codegen/cuda/arith.cpp

```diff
@@ -1175,16 +1274,87 @@ TensorView* sum_to(TensorView* in, const std::vector<int64_t>& sum_to_size) {
   return out;
 }
 
-TensorView* shift(TensorView* inp, const std::vector<int>& offsets) {
+TensorView* shift(TensorView* inp, const std::vector<int>& offsets, bool pad) {
   TORCH_CHECK(
       TensorDomain::noReductions(inp->getRootDomain()).size() == offsets.size(),
       "Invalid shift offsets, number of entries in offsets expected to be ",
       TensorDomain::noReductions(inp->getRootDomain()).size(),
       " but received ",
       offsets.size());
 
-  auto out = newValLike(inp, inp->getDataType().value())->as<TensorView>();
-  new ShiftOp(out, inp, offsets);
+  TensorView* out = nullptr;
+
+  if (pad) {
+    out = newValLike(inp, inp->getDataType().value())->as<TensorView>();
+  } else {
+    auto inp_dom = TensorDomain::noReductions(inp->getRootDomain());
+    const auto ndims = inp_dom.size();
+    std::vector<IterDomain*> out_dom;
+    for (size_t i = 0; i < ndims; ++i) {
+      const auto inp_axis = inp_dom[i];
+      const auto offset = offsets[i];
+      if (offset == 0) {
+        out_dom.push_back(inp_axis->clone());
+        continue;
+      }
+
+      Int* current_start = dynamic_cast<Int*>(inp_axis->start());
+      TORCH_INTERNAL_ASSERT(
+          current_start != nullptr && current_start->isConst(),
+          "Invalid IterDomain start value:",
+          current_start);
+
+      const auto cur_start_offset = current_start->value().value();
+      const auto cur_stop_offset = getStopOffset(inp_axis);
+
+      Val* start = nullptr;
+      Val* stop = nullptr;
+
+      if (offset > 0) {
+        // shift to right; extent remains the same, start and stop
+        // positions are moved right
```

Comment:
Positive offsets move tensors right. Maybe "offsets" isn't a good name. It specifies a direction to shift.

The reason that the stop position is also adjusted is because it may not be equal to the extent. Since this is to shift to right, the stop position is also shifted right, i.e., `stop + offset`, and the `max` expression is because since `stop` never gets larger than `extent`. These adjustment computations are done statically when possible, but if not, it uses Fusion expressions to represent the new stop value.